src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
-    type Err = DeltaParserError;
+    /// Marks a snapshot file as deprecated. Assumes that the session id is still unchanged. If the
+    /// there would be a session id reset, then deprecate files for the old session id first, before
+    /// updating the current session id.
+    fn deprecate_snapshot_file(&mut self) {
+        let path = Self::path_snapshot(&self.rrdp_dir, self.session_id(), self.serial());
+        self.deprecated_files.push(DeprecatedFile::new(path));
+    }
+
+    /// Writes a new notification file. Will not check whether the file already
+    /// exists because this is assumed to be called for new snapshot files only.
+    fn write_snapshot(&self) -> Result<()> {
+        let path = Self::path_snapshot(&self.rrdp_dir, self.session_id(), self.serial());
+        if path.exists() {
+            debug!("Skip writing existing snapshot file to {:?}", path)
+        } else {
+            info!("Writing snapshot file to {:?}", path);
+            let xml = self
+                .snapshot
+                .xml()
+                .ok_or_else(|| anyhow!("Snapshot XML not recovered on startup"))?;
+            file_ops::write_buf(&path, xml).with_context(|| "Could not write snapshot XML")?;
+        }
 
-    fn meta(&mut self, session_id: Uuid, serial: u64) -> Result<(), Self::Err> {
-        self.meta = Some(ParsedMeta { session_id, serial });
         Ok(())
     }
 
-    // add the delta to the snapshot
-    fn publish(&mut self, uri: uri::Rsync, _hash: Option<DigestHex>, data: Vec<u8>) -> Result<(), Self::Err> {
-        if self.is_applicable()? {
-            // add it to the snapshot, don't check if it already existed because
-            // by URI it may already exist but just with a different value, which
-            // it is correct that we overwrite
-            self.snapshot.publish.insert(ComparableRsyncUri(uri), data);
+    /// Marks a delta file as deprecated. Assumes that the session id is still unchanged. If the
+    /// there would be a session id reset, then deprecate files for the old session id first, before
+    /// updating the current session id.
+    fn deprecate_delta_file(&mut self, serial: u64) {
+        let path = Self::path_delta(&self.rrdp_dir, self.session_id(), serial);
+        self.deprecated_files.push(DeprecatedFile::new(path));
+    }
+
+    /// Writes any deltas for which no current file is found. I.e. it is assumed
+    /// that a file which is present was not tampered with since writing it.
+    fn write_missing_deltas(&self) -> Result<()> {
+        for delta in &self.deltas {
+            let path = Self::path_delta(&self.rrdp_dir, self.session_id(), delta.serial);
+            if path.exists() {
+                debug!("Skip writing delta file to {:?}", path)
+            } else {
+                info!("Writing delta file to {:?}", path);
+                let xml = delta
+                    .xml()
+                    .ok_or_else(|| anyhow!("Delta XML not recovered on startup"))?;
```

Comment:
Should the error context include the path of the XML file?

Reply:
done